src/amo/components/UserProfileEdit/index.js

```diff
@@ -0,0 +1,475 @@
+/* @flow */
+import invariant from 'invariant';
+import * as React from 'react';
+import Textarea from 'react-textarea-autosize';
+import Helmet from 'react-helmet';
+import { connect } from 'react-redux';
+import { compose } from 'redux';
+
+import Link from 'amo/components/Link';
+import NotFound from 'amo/components/ErrorPage/NotFound';
+import {
+  editUserAccount,
+  fetchUserAccount,
+  getCurrentUser,
+  getUserByUsername,
+  hasPermission,
+} from 'amo/reducers/users';
+import { USERS_EDIT } from 'core/constants';
+import { withErrorHandler } from 'core/errorHandler';
+import translate from 'core/i18n/translate';
+import { sanitizeHTML } from 'core/utils';
+import Button from 'ui/components/Button';
+import Card from 'ui/components/Card';
+import Notice from 'ui/components/Notice';
+import type { UsersStateType, UserType } from 'amo/reducers/users';
+import type { DispatchFunc } from 'core/types/redux';
+import type { ErrorHandlerType } from 'core/errorHandler';
+import type { I18nType } from 'core/types/i18n';
+
+import './styles.scss';
+
+
+type Props = {|
+  currentUser: UserType | null,
+  dispatch: DispatchFunc,
+  errorHandler: ErrorHandlerType,
+  hasEditPermission: boolean,
+  i18n: I18nType,
+  isEditing: boolean,
+  // The routing `params` prop is used in `mapStateToProps()`.
+  // eslint-disable-next-line react/no-unused-prop-types
+  params: {| username: string |},
+  user: UserType | null,
+  username: string,
+|};
+
+type FormValues = {|
+  biography: string | null,
+  displayName: string | null,
+  homepage: string | null,
+  location: string | null,
+  occupation: string | null,
+  username: string | null,
+|};
+
+type State = {|
+  ...FormValues,
+  displaySuccessMessage: boolean,
+|};
+
+export class UserProfileEditBase extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      displaySuccessMessage: false,
+      ...this.getFormValues(props.user),
+    };
+  }
+
+  componentWillMount() {
+    const { dispatch, errorHandler, username, user } = this.props;
+
+    if (!user && username) {
+      dispatch(fetchUserAccount({
+        errorHandlerId: errorHandler.id,
+        username,
+      }));
+    }
+  }
+
+  componentWillReceiveProps(props: Props) {
+    const { isEditing: wasEditing, username: oldUsername } = this.props;
+    const {
+      dispatch,
+      errorHandler,
+      isEditing,
+      username: newUsername,
+    } = props;
+
+    if (oldUsername !== newUsername) {
```

Comment:
I know you changed this because I felt the previous code was confusing, but I think there's a problem here now. It won't properly update the state if the username does change. You can see this if you run the code and update your username, and click "Update", you get a screen with the old username populated and all the other fields empty.

I think it's missing the ability to handle receiving the new user data after `fetchUserAccount` returns the data.